### PR TITLE
User level dashboard theming (dashboard banners)

### PR DESCRIPTION
**Pull request by lasanthaS**

### CODE REVIEW

File: apps/portal/controllers/apis/banners.jag [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            try {
+                
+                // decode the base64 string
+                var decodedData = Packages.javax.xml.bind.DatatypeConverter.parseBase64Binary(encodedData);
+
+                // save the buffer in the registry
+                dashboards.saveBanner(dashboardId, username, 'banner', 'image/jpg', decodedData); 
+                
+            } catch(e) {
+                log.error('Error occurred while saving the banner');
```
Comment (TanyaM) [inline apps/portal/controllers/apis/banners.jag]: can do in one line as log.error(message, throwable);

File: apps/portal/controllers/apis/banners.jag [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+                log.error(e);
```
Comment (TanyaM) [inline apps/portal/controllers/apis/banners.jag]: can do in one line as log.error(message, throwable);

File: apps/portal/js/designer.js [hunk 3]
```diff
@@ -1954,9 +1963,165 @@ $(function () {
         }
         renderPage(page || db.landing || pages[0].id);
     };
+    
+    /**
+     * load the content within the banner placeholder
+     */
+    var loadBanner = function() {        
+        
+        ues.global.dashboard.banner = ues.global.dashboard.banner || { globalBannerExists: false, customBannerExists: false };
```
Comment (TanyaM) [inline apps/portal/js/designer.js]: please check line length is set to 120

File: apps/portal/controllers/apis/banners.jag [hunk 4]
```diff
@@ -0,0 +1,87 @@
+<%
+/*
+ * Copyright 2005-2015 WSO2, Inc. (http://wso2.com)
```
Comment (TanyaM) [inline apps/portal/controllers/apis/banners.jag]: use the correct license

File: apps/portal/controllers/banners.jag [hunk 5]
```diff
@@ -0,0 +1,51 @@
+<%
+/*
+ * Copyright 2005-2015 WSO2, Inc. (http://wso2.com)
```
Comment (TanyaM) [inline apps/portal/controllers/banners.jag]: License

File: apps/portal/modules/dashboards.js [hunk 6]
```diff
@@ -3,6 +3,16 @@ var log = new Log();
 var carbon = require('carbon');
 var utils = require('/modules/utils.js');
 
+/**
+ * get registry reference 
+ */
```
Comment (TanyaM) [inline apps/portal/modules/dashboards.js]: Add return type

File: apps/portal/modules/dashboards.js [hunk 7]
```diff
@@ -152,3 +151,113 @@ var allowed = function (dashboard, permission) {
...
+    if (registry.exists(path)) {
+        
+        result.customBannerExists = true;
+        result.path = path;
+    }
+    
+    // check to see if there is any global banner
+    path = registryBannerPath(dashboardId, null);
+    if (registry.exists(path)) {
+
```
Comment (TanyaM) [inline apps/portal/modules/dashboards.js]: remove new line

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs [hunk 8]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+
```
Comment (TanyaM) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: new line

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs [hunk 9]
```diff
@@ -0,0 +1,21 @@
...
+    <div class="row">
+        <div id="d" class="col-md-6 ues-component-box"></div>
+        <div id="e" class="col-md-6 ues-component-box"></div>
+    </div>
+    <div class="row">
+        <div id="f" class="col-md-4 ues-component-box"></div>
+        <div id="g" class="col-md-4 ues-component-box"></div>
+        <div id="h" class="col-md-4 ues-component-box"></div>
+    </div>
+
```
Comment (TanyaM) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: new line

File: apps/portal/theme/templates/dashboard-settings.jag [hunk 10]
```diff
@@ -24,8 +24,8 @@
                         <i class="fw fw-ring fw-stack-2x"></i>
                         <i class="fw fw-left-arrow fw-stack-1x"></i>
                     </span> <%= i18n.localize("go.back")%></a></li>
-                <li><a href="<%=tenantedUrlPrefix%>dashboards/<%=dashboardId%>?editor=true">
-                <span class="fw-stack">
+                <li><a href="<%=tenantedUrlPrefix%>dashboards/<%=dashboardId%>?editor=true" id="ues-designer-link">
```
Comment (TanyaM) [inline apps/portal/theme/templates/dashboard-settings.jag]: format code. <li> and <a> not aligned.

File: apps/portal/theme/templates/designer.jag [hunk 11]
```diff
@@ -195,6 +198,7 @@
 <script src="<%=urlPrefix%>js/jquery-1.10.2.js"></script>
 <script src="<%=urlPrefix%>js/bootstrap.min.js"></script>
 <script src="<%=urlPrefix%>js/jquery-ui.js"></script>
+<script src="<%=urlPrefix%>js/cropper.js"></script>
```
Comment (TanyaM) [inline apps/portal/theme/templates/designer.jag]: Move the scripts that can be moved to bottom of the page to speed up the page loading.

File: apps/portal/theme/templates/designer.jag [hunk 12]
```diff
@@ -490,5 +493,43 @@
...
+    
+    <canvas id="src-canvas" class="hidden"></canvas>    
+    <canvas id="dest-canvas" class="hidden"></canvas>
+    
+    {{#if isEdit}}
+    <img src="" id="img-banner" data-src="<%=urlPrefix%>banners/<%=dashboard.id%>" style="display: none;" />
+    {{/if}}
+       
+    <div class="button-panel">
+
```
Comment (TanyaM) [inline apps/portal/theme/templates/designer.jag]: new line

File: apps/portal/theme/templates/designer.jag [hunk 13]
```diff
@@ -490,5 +493,43 @@
...
+    {{/if}}
+       
+    <div class="button-panel">
+    
+        {{#if isAdd}}
+        <button type="button" id="btn-edit-banner" class="btn btn-primary">Add</button>
+        {{/if}}
+
+        {{#if isEdit}}
+        <button type="button" id="btn-edit-banner" class="btn btn-primary">Edit</button>
```
Comment (TanyaM) [inline apps/portal/theme/templates/designer.jag]: Use localization for button names

File: apps/portal/theme/templates/designer.jag [hunk 14]
```diff
@@ -490,5 +493,43 @@
...
+        {{#if showRemove}}
+        <button type="button" id="btn-remove-banner" class="btn btn-danger">{{#if isCustomBanner}}Reset{{else}}Remove{{/if}}</button>
+        {{/if}}
+        {{/if}}
+
+        {{#if isCrop}}
+        <button type="button" id="btn-save-banner" class="btn btn-primary">Save</button>
```
Comment (TanyaM) [inline apps/portal/theme/templates/designer.jag]: localization

File: apps/portal/theme/templates/designer.jag [hunk 15]
```diff
@@ -490,5 +493,43 @@
...
+        {{/if}}
+    
+    </div>
+    
+</script> <!--/#ues-dashboard-banner-hbs -->
+    
+<form method="post" id="ues-dashboard-upload-banner-form" action="<%=urlPrefix%>apis/banners/<%=dashboard.id%>" class="hidden">
+    <input type="file" id="file-banner" name="banner" accept="image/*" />
+    <input type="hidden" id="banner-data" name="data" />
+    <button type="submit">Submit</button>
```
Comment (TanyaM) [inline apps/portal/theme/templates/designer.jag]: Localization

File: apps/portal/js/designer.js [hunk 16]
```diff
@@ -1954,9 +1963,165 @@ $(function () {
...
+            isAdd: !bannerExists && !banner.cropMode,
+            isEdit: bannerExists && !banner.cropMode,
+            isCrop: banner.cropMode,
+            isCustomBanner: customDashboard && banner.customBannerExists,
+            showRemove: (customDashboard && banner.customBannerExists) || !customDashboard
+        };
+        $placeholder.html(bannerHbs(data));
+        
+        // display the image 
+        var img = $placeholder.find('#img-banner');
```
Comment (TanyaM) [inline apps/portal/js/designer.js]: Refactor the varible name specifying which image

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs [hunk 17]
```diff
@@ -0,0 +1,21 @@
+<div class="container">
+
+   <div class="row">
+        <div class="col-md-12 ues-banner-placeholder"></div>
+   </div>
+   <div class="row">
+        <div id="a" class="col-md-3 ues-component-box"></div>
```
Comment (TanyaM) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: Specify other bootstrap classes

File: apps/portal/theme/templates/dashboard.jag [hunk 18]
```diff
@@ -198,6 +198,7 @@
 <!--<iframe id="sandbox" src="preview.jag"></iframe>-->
```
Comment (TanyaM) [inline apps/portal/theme/templates/dashboard.jag]: Remove comments

File: apps/portal/css/cropper.css [hunk 19]
```diff
@@ -0,0 +1,290 @@
+/*!
+ * Cropper v1.0.0-rc.1
```
Comment (TanyaM) [inline apps/portal/css/cropper.css]: has this library been approaved at stratergy@

File: apps/portal/modules/dashboards.js [hunk 20]
```diff
@@ -70,16 +78,15 @@ var create = function (dashboard) {
...
-    var registry = new carbon.registry.Registry(server, {
-        system: true
-    });
+    
+    var registry = getRegistry();
 
     var usr = require('/modules/user.js');
     var user = usr.current();
     if(!user){
         throw 'User is not logged in ';
```
Comment (TanyaM) [inline apps/portal/modules/dashboards.js]: Do we have a catch block in the call stack?

File: apps/portal/controllers/apis/banners.jag (hunk 4 above)
Comment (lasanthaS) [inline apps/portal/controllers/apis/banners.jag]: Fixed in the PR: https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/controllers/apis/banners.jag (hunk 2 above)
Comment (lasanthaS) [inline apps/portal/controllers/apis/banners.jag]: Fixed in the PR: https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/controllers/banners.jag (hunk 5 above)
Comment (lasanthaS) [inline apps/portal/controllers/banners.jag]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/js/designer.js (hunk 3 above)
Comment (lasanthaS) [inline apps/portal/js/designer.js]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/js/designer.js (hunk 16 above)
Comment (lasanthaS) [inline apps/portal/js/designer.js]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/modules/dashboards.js (hunk 6 above)
Comment (lasanthaS) [inline apps/portal/modules/dashboards.js]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/modules/dashboards.js (hunk 7 above)
Comment (lasanthaS) [inline apps/portal/modules/dashboards.js]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs (hunk 8 above)
Comment (lasanthaS) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: This file is no longer available

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs (hunk 17 above)
Comment (lasanthaS) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: This file is no longer available

File: apps/portal/store/carbon.super/layout/layout-banner/index.hbs (hunk 9 above)
Comment (lasanthaS) [inline apps/portal/store/carbon.super/layout/layout-banner/index.hbs]: This file is no longer available

File: apps/portal/theme/templates/dashboard-settings.jag (hunk 10 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/dashboard-settings.jag]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/theme/templates/designer.jag (hunk 15 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/designer.jag]: This is already fixed

File: apps/portal/theme/templates/designer.jag (hunk 14 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/designer.jag]: This is already fixed

File: apps/portal/theme/templates/designer.jag (hunk 11 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/designer.jag]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/theme/templates/designer.jag (hunk 13 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/designer.jag]: This is already fixed

File: apps/portal/theme/templates/dashboard.jag (hunk 18 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/dashboard.jag]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

File: apps/portal/theme/templates/designer.jag (hunk 12 above)
Comment (lasanthaS) [inline apps/portal/theme/templates/designer.jag]: Fixed in the PR https://github.com/wso2/carbon-dashboards/pull/174

**Pull request merged**